paasta_tools/kubernetes/bin/kubernetes_remove_evicted_pods.py

```diff
@@ -0,0 +1,155 @@
+from paasta_tools.utils import DEFAULT_SOA_DIR
+
+
+log = logging.getLogger(__name__)
+EvictedPod = namedtuple("EvictedPod", ["podname", "namespace", "eviction_msg"])
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        description="Removes evicted pods and notifies service owners"
+    )
+    parser.add_argument(
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+    parser.add_argument(
+        "-v", "--verbose", action="store_true", dest="verbose", default=False
+    )
+    parser.add_argument(
+        "-n", "--dry-run", action="store_true", dest="dry_run", default=False
+    )
+    args = parser.parse_args()
+    return args
+
+
+def get_evicted_pods(pods: Sequence[V1Pod]) -> Sequence[V1Pod]:
+    return [
+        pod
+        for pod in pods
+        if pod.status.phase == "Failed" and pod.status.reason == "Evicted"
+    ]
+
+
+def get_pod_service(pod: V1Pod) -> str:
+    return pod.metadata.labels.get("paasta.yelp.com/service")
+
+
+def notify_service_owners(
+    services: Mapping[str, List[EvictedPod]], soa_dir: str, dry_run: bool
+) -> None:
+    check_overrides = {
+        "page": False,
+        "alert_after": "0m",
+        "realert_every": 1,
+        "tip": "Pods can be Evicted if they go over the allowed quota for a given resource. Check the Eviction message to figure out which resource quota was breached",
+    }
+    for service in services.keys():
+        check_name = f"pod-eviction.{service}"
+        check_output = "The following pods have been evicted and will be removed from the cluster:\n"
+        for pod in services[service]:
+            check_output += f"- {pod.podname}: {pod.eviction_msg}\n"
+        if dry_run:
+            log.info(f"Would have notified owners for service {service}")
+        else:
+            log.info(f"Notifying owners for service {service}")
+            send_event(
+                service,
+                check_name,
+                check_overrides,
+                Status.CRITICAL,
+                check_output,
+                soa_dir,
+            )
+
+
+def remove_pods(
+    client: KubeClient, services: Mapping[str, List[EvictedPod]], dry_run: bool
+) -> None:
+    delete_options = V1DeleteOptions()
+    for service in services:
+        # Do not remove more than 2 pods per run
+        for pod in services[service][0:2]:
+            if dry_run:
+                log.info(f"Would have removed pod {pod.podname}")
+            else:
+                client.core.delete_namespaced_pod(
+                    pod.podname,
+                    pod.namespace,
+                    body=delete_options,
+                    grace_period_seconds=0,
+                )
+                log.info(f"Removing pod {pod.podname}")
+
+
+def evicted_pods_per_service(client: KubeClient) -> Mapping[str, List[EvictedPod]]:
```

Comment:
nit: it's preferable to use `Sequence` or `Iterable` (depending on do you need indexable access or not) instead of `List` of the list is read-only.

Reply:
I noticed that **mypy** is throwing: `Need type annotation for 'evicted_pods_aggregated'`.

**evicted_pods_aggregated** is used on the return statement, so its type needs to match the type on the function signature (otherwise mypy fails).

So the best I can do is to switch from **List** to  **MutableSequence** (unless I'm missing something else here)

Reply:
The type of `evicted_pods_aggregated` needs to be a subtype of the return type for `evicted_pods_per_service()` if the former is returned from the later.
So, the former could be e.g. `Dict[std, List[EvictedPod]]`, while the later could be `Mapping[str, Iterable[EvictedPod]]`.

Reply:
```python
#! /usr/bin/env python

from typing import *

def f() -> Mapping[str, Iterable[int]]:
    z : Dict[str, List[int]] = {}
    z["qqq"] = [1,2,3]
    z["qqq"].append(4)
    return z

print(f())
```
```bash
$ ./1.py
{'qqq': [1, 2, 3, 4]}

$ mypy --strict 1.py
Success: no issues found in 1 source file
```

Reply:
That makes a lot of sense, thanks for the pointer! I'm going to update the PR in a moment.
By the way, I'll be using Sequence instead of Iterable, as I need index access on one of the functions.